### PR TITLE
Views: Support contributing welcome-views to empty tree view

### PR DESCRIPTION
**Pull request by EstherPerelman**
Signed-off-by: Esther Perelman <esther.perelman@sap.com>

Welcome-views can be contributed by extensions in order to be applied to empty tree views, A view is considered empty if the tree has no children.
see vscode doc: [vs-code-contributes.viewsWelcome](https://code.visualstudio.com/api/references/contribution-points#contributes.viewsWelcome)
#### What it does
Partially solves #7178  #8667  (For now supports contributing to custom-views and 'explorer' view but not the other built-in views: 'scm', 'debugger', 'test').
#### How to test
1. Deploy test extension [run-configuration.zip](https://github.com/eclipse-theia/theia/files/5452263/run-configuration.zip)
When opening it from the left menu - you should see the contributed welcome-view:
Screenshot 2020-10-28 155722
2. You can now contribute more welcome-views to the run-configuration view 
by adding a viewWelcome to any extension package.json under "contributes" key, For example:

   "viewsWelcome": [{
       "view": runConfigurations,
       "contents": "In order to use git features, you can open a folder containing a git repository or clone from a URL.\n[Open Folder](command:vscode.openFolder)\n[Clone Repository](command:git.clone)\nTo learn more about how to use git and source control in VS Code [read our docs](https://aka.ms/vscode-scm).",
      }]

Now it should look like this:
Screenshot 2020-10-28 161547

You can play around... (add when clause etc...)

3. open theia without opening a workspace/folder - you should see a new welcome view in the Explorer tab, You can contribute additional views-welcome to the Explorer view by specifying "view": "explorer" .

The contentns are rendered like this: 
'/n' - for new line.
'[here goes the label](here goes a command or a link)' - If there's only a command/link in a line without any other text - it will be displayed as a button, Otherwise as a link.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/main/browser/style/tree.css
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/style/tree.css:49]: We should not use hardcoded constants but rather use variables.
The same applies to other css updates included in the pull-request.

File: packages/plugin-ext/src/main/browser/style/tree.css
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/style/tree.css:66]: Please fix the formatting inconsistencies.

File: packages/plugin-ext/src/main/browser/view/plugin-view-welcome.tsx
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-welcome.tsx:2]: 2020 SAP?

**Review by vince-fugnitto — COMMENTED** [review]
We can perform a comprehensive review after the next release (tomorrow) :+1:

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-view-welcome-widget.tsx:77]: `<div className="theia-WelcomeView">` -> `<div className='theia-WelcomeView'>`

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:77]: Please change other double quotes to single: `"` -> `'`

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-view-welcome-widget.tsx:212]: This must go through the `WindowService`. Otherwise, it can be a security issue in Electron.

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-view-welcome-widget.tsx:219]: It is very important to open a CQ if you have copied code from an open-source project.

https://github.com/microsoft/vscode/blob/573e5145ae3b50523925a6f6315d373e649d1b06/src/vs/base/common/linkedText.ts#L26

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-view-welcome-widget.tsx:40]: This PR requires a CQ for sure:

https://github.com/microsoft/vscode/blob/573e5145ae3b50523925a6f6315d373e649d1b06/src/vs/base/common/linkedText.ts#L8-L12


@marcdumais-work, @amiramw can one of you please help @EstherPerelman with the CQ submission? The changeset works pretty great. Theia needs this addition but we have to clearly document, the idea was taken from VS Code. Thank you!

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-view-welcome-widget.tsx:55]: Generally, it's better to have `protected` everywhere instead of `private`. Just think about it, what if you would like to quickly customize something on your side, `protected` visibility could come in handy. Please change. Thanks!

File: packages/filesystem/src/browser/file-tree/file-tree-widget.tsx
Comment (kittaakos) [inline packages/filesystem/src/browser/file-tree/file-tree-widget.tsx:36]: Can we enable the welcome-support in all other trees as well? It isn't FS specific. What do you think, @akosyakov?

**Review by kittaakos — COMMENTED** [review]
I tried it with a VS Code extension, and it works perfectly. I look forward to seeing this feature in Theia. Nice job, @EstherPerelman  👍

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (amiramw) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @kittaakos i have never dealt with CQ. is the process documented somewhere?

**Comment by EstherPerelman** [comment]
@kittaakos @vince-fugnitto Can you continue reviewing the code?

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (kittaakos) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: Yes, it is documented on the Wiki. [Here](https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-the-main-vs-code-repo) is the relevant section when the code was copied from VS Code.

File: packages/search-in-workspace/package.json
Comment (DukeNgn) [inline packages/search-in-workspace/package.json:12]: I believe you made this change by mistake when rebasing. Please revert it as `minimatch` is required for siw.

File: packages/core/src/browser/tree/tree-view-welcome-widget.tsx
Comment (marcdumais-work) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @amiramw have you opened the CQ? With the increased contributions, it's good that one SAP committer joins the ranks of our "IP champions". Let me know if I have questions. 

If by chance you can find the code you want, already covered by a previous, approved CQ, then we would not need a new one. For example, we often request to copy from `vscode.d.ts`.

Comment (amiramw) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @marcdumais as i understand since this copied code is from newer version then i need to open a CQ based on https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-other-projectsrepos, right?

Where can I see all previous CQs?

Comment (amiramw) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @kittaakos maybe you can answer?

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: > Where can I see all previous CQs?

No idea. I bet filtering by label does not show all: https://github.com/eclipse-theia/theia/pulls?q=+is%3Apr+label%3A%22CQ+Required%22+

Comment (marcdumais-work) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: > @marcdumais as i understand since this copied code is from newer version then i need to open a CQ based on https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-other-projectsrepos, right?
> 

Yes, I think so. The exception being if the code has been there for a while, in a form you can use, then you are allowed to pick the code from whatever version you want.

> Where can I see all previous CQs?

I am no expert, but I start [here](https://dev.eclipse.org/ipzilla/query.cgi) and usually only do a simple keywork search, e.g. for [vscode](https://dev.eclipse.org/ipzilla/buglist.cgi?query_format=advanced&short_desc_type=allwordssubstr&short_desc=vscode&long_desc_type=substring&long_desc=&bug_file_loc_type=allwordssubstr&bug_file_loc=&keywords_type=allwords&keywords=&emailassigned_to1=1&emailtype1=substring&email1=&emailassigned_to2=1&emailreporter2=1&emailcc2=1&emailtype2=substring&email2=&bugidtype=include&bug_id=&chfieldfrom=&chfieldto=Now&chfieldvalue=&cmdtype=doit&order=Reuse+same+sort+as+last+time&field0-0-0=noop&type0-0-0=noop&value0-0-0=) or whatever I am looking for.

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: Has anyone already registered the CQ? I do not want to hold back this great change. So if this PR is blocked only by me, let me know.

Comment (amiramw) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @kittaakos @marcdumais-work I registered CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22793 now. It's my first time so I hope someone can take a look at it.

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: Great! Thank you so much.

I looked into the attachment, and it looks good to me. I would have created the same archive.🤞

Comment (amiramw) [reply packages/core/src/browser/tree/tree-view-welcome-widget.tsx:47]: @kittaakos the CQ is approved now

**Comment by amiramw** [comment]
@kittaakos @vince-fugnitto other then the CQ do you have more comments?

**Review by vince-fugnitto — COMMENTED** [review]
The changes look good to me, I verified with the simple extension but not the additional configuration as I do not have the source code.

I'm wondering about the following https://github.com/eclipse-theia/theia/pull/8678/files#r515101313 however. I'll add reviewers who are more familiar with the plugin-system to see if they have any comments as well.

**Review by kittaakos — COMMENTED** [review]
It works nicely, very good job. I added a few comments.

One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?

Screen Shot 2020-11-20 at 10 08 08

**Comment by EstherPerelman** [comment]
> It works nicely, very good job. I added a few comments.
> 
> One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> 
> Screen Shot 2020-11-20 at 10 08 08

@amiramw What do you think?

File: packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts
Comment (EstherPerelman) [reply packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts:522]: I just got Eslint error:
The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype
[.eslintguard-for-in](https://eslint.org/docs/rules/guard-for-in)
Should I disable it?

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts
Comment (EstherPerelman) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:142]: Sorry..
But what should be the text there? any idea?

**Comment by amiramw** [comment]
> > It works nicely, very good job. I added a few comments.
> > One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> > Screen Shot 2020-11-20 at 10 08 08
> 
> @amiramw What do you think?

It should be supported when there is no opened workspace. At least in vscode it is with vscode builtin git support. Did you test with builtin git extension?

File: packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts
Comment (amiramw) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:142]: I'm not sure we have to register this view welcome for explorer. at some point we can just move to builtin git extension that configure it anyway, no?

Comment (EstherPerelman) [reply packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts:142]: The second section will be removed and contribute by vsCode builtin git extension.

**Comment by EstherPerelman** [comment]
> It works nicely, very good job. I added a few comments.
> 
> One question; if _Clone Repository_ is not supported when there is no opened workspace, why do we show it?
> 
> Screen Shot 2020-11-20 at 10 08 08

@kittaakos Thanks for reviewing! I updated the code accordingly

**Comment by kittaakos** [comment]
> It should be supported when there is no opened workspace

There was no workspace opened:
Screen Shot 2020-11-23 at 13 34 41


> Did you test with builtin git extension?

I tested with the electron example app.

**Comment by EstherPerelman** [comment]
> > It should be supported when there is no opened workspace
> 
> There was no workspace opened:
> Screen Shot 2020-11-23 at 13 34 41
> 
> > Did you test with builtin git extension?
> 
> I tested with the electron example app.

Anyway, I removed it and when working with the builtin git extension - it will be contributed by it (and supported).

**Comment by amiramw** [comment]
@kittaakos are there any more comments or can we merge?

**Comment by kittaakos** [comment]
> are there any more comments or can we merge?

I haven't had time to look into the updates, probably it is good to merge if all remarks have been addressed, but we need to restore the CI first.

**Comment by amiramw** [comment]
@kittaakos PR is rebased and build passed. could you give final approval?

**Review by kittaakos — APPROVED** [review]
Thank you, @EstherPerelman. Nice feature👍 

@amiramw, please help with the merging. I've checked and tried the changeset twice already.

**Review by amiramw — APPROVED** [review]
Went over the code again and tested successfully

**Pull request merged**
